### PR TITLE
VariableRateShading RHI sample

### PR DESCRIPTION
**Pull request by akioCL**
Fixes [#13388](https://github.com/o3de/o3de/issues/13388)

Signed-off-by: Akio Gaule <10719597+akioCL@users.noreply.github.com>

Sample for the VRS implementation of the RHI.
image

image

image

image

[RFC ](https://github.com/o3de/sig-graphics-audio/issues/87 )for VRS 

[PR](https://github.com/o3de/o3de/pull/13389) for the VRS implementation

### CODE REVIEW

File: Gem/Code/Source/RHI/VariableRateShadingExampleComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,874 @@
...
+        m_computeShaderResourceGroup = CreateShaderResourceGroup(m_shaders[1], "ComputeSrg", VariableRateShading::SampleName);
+        FindShaderInputIndex(&patternIndex, m_computeShaderResourceGroup, patternId, VariableRateShading::SampleName);
+        FindShaderInputIndex(&m_centerIndex, m_computeShaderResourceGroup, centerId, VariableRateShading::SampleName);
+        FindShaderInputIndex(&m_shadingRateIndex, m_computeShaderResourceGroup, shadingRateImageId, VariableRateShading::SampleName);
+
+        RHI::Ptr<RHI::Device> device = Utils::GetRHIDevice();
+
+        struct Pattern
+        {
+            float m_distance[4];
```
Comment (moudgils) [inline Gem/Code/Source/RHI/VariableRateShadingExampleComponent.cpp:288]: pull out all the 4s in a variable

**Review by jeremyong-az — APPROVED** [review]

File: Gem/Code/Source/RHI/VariableRateShadingExampleComponent.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    //! When a PerDraw mode is used, the rate is applied equally to the whole quad. The rate can be changed
+    //! using the GUI of the sample.
+    //! Combinator operations are also exposed when both PerDraw and PerRegion are being used.
+    class VariableRateShadingExampleComponent final
+        : public BasicRHIComponent
+        , public AZ::TickBus::Handler
+        , public AzFramework::InputChannelEventListener
+    {
+    public:
+        AZ_COMPONENT(VariableRateShadingExampleComponent, "{B98E1C6A-8C23-4AA4-82E6-4B652F6151DD}", AZ::Component);
```
Comment (moraaar) [inline Gem/Code/Source/RHI/VariableRateShadingExampleComponent.h:53]: I think the third parameter of `AZ_COMPONENT` if the base class, which in this case is `BasicRHIComponent`.

File: Shaders/RHI/VariableRateShading.azsl [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Atom/Features/SrgSemantics.azsli>
+ShaderResourceGroup InstanceSrg : SRG_PerObject
```
Comment (moraaar) [inline Shaders/RHI/VariableRateShading.azsl:10]: This is probably be a silly question, but I noticed there are no shaders involved in the [o3de PR](https://github.com/o3de/o3de/pull/13389), doesn't this technique require any shaders in the  engine side?

**Review by moraaar — APPROVED** [review]
I did a quick pass on this. LGTM. Just a couple of comments.

**Pull request merged**